<a href="https://colab.research.google.com/github/ChronosXV/Ingenieria_Software/blob/Ing_Software/Proyecto/UGR_Ing_de_Software_Grupo_2_TP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Universidad del Gran Rosario - UGR
#Ingeniería del Software
##Trabajo Práctico
###Introducción a las metodologías y procesos de Ciencia de Datos
####Docentes: Ing. Ignacio Sanseovich - Lic. Briant Gauna

####Estudiantes:
*   Gisela Martinez
*   Noelia Cualina
*   Javier Carabajal
*   Leonardo Carabajal
*   Nicolas Suarez
*   Marcelo Delgado
*   Ezequias Bonvissuto
*   Pedro Roberti


###Estudio de Caso
Este trabajo se basa en la aplicación de una metodología de Minería de Datos y su implementación en una solución de software.

Lo que se espera es que el grupo desarrolle:
* Una metodología clara de cómo se procesarán los datos y su resultado e interpretación.
* Las vistas y montaje de la solución de software que dará vida a esta metodología para conectar el proceso de datos con el usuario final.

Elegiremos trabajar con la Metodología CRISP-DM (Cross-Industry Standard Process for Data Mining).

La entrega constará de:
* Esta notebook completa
* El link desde donde se puede acceder al producto de software implementado

##Fases del Trabajo:

##Fase #1: Comprensión del Negocio / Business Understanding:
Esta fase inicial se centra en comprender los objetivos y requisitos del proyecto desde una perspectiva de negocio, para luego convertir este conocimiento en una definición del problema de minería de datos y un plan preliminar diseñado para lograr los objetivos.

* Elegí un problema de negocio que puedas resolver con datos. Describilo brevemente.
* ¿Cuál es el objetivo principal del análisis? Elaborá un objetivo general y algunos específicos. Te ayudarán a clasificar las tareas más adelante.
* Enumerá los distintos Stakeholders del negocio donde está inmerso el problema.
* ¿Qué impacto tendrá la solución en el negocio o en la toma de decisiones? Justificála realización del estudio desde la perspectiva de cada Stakeholder enumerado anteriormente.

##Fase #2: Comprensión de los Datos / Data Understanding
La fase de comprensión de datos comienza con una recopilación inicial de datos y continúa con actividades para familiarizarse con los datos, identificar problemas de calidad de los datos, descubrir los primeros conocimientos sobre los datos o detectar subconjuntos interesantes para formular hipótesis sobre información oculta.
* Establecé una metodología de adquisición de los datos para el negocio.
* Contemplá en la adquisición de datos una alternativa donde debas construir el set de datos.
* Identificá los distintos problemas o desafíos con los datos (datos faltantes, ruidos,etc) que puedan afectar a su calidad.
* Enumerá las distintas fases, tareas generales, tareas específicas e instancias del proceso que incluirás como estrategia para aplicar la metodología

# Librerias importadas

In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from google.colab import userdata
import plotly.express as px



## Inicializacion del repositorio GIT

In [2]:
# Seteo de variables para seteo de config globals para git
git_email: str = userdata.get('git_email')
git_name: str = userdata.get('git_name')
git_token: str = userdata.get('git_token')

In [3]:
# Seleccion de branch por integrantes del equipo
branch_name_ing: str = userdata.get('branch_name_ing')

In [4]:
# Setear la config global de git
!git config --global user.email {git_email}
!git config --global user.name {git_name}

In [5]:
# Ejecutamos el comando git clone y descargamos nuestro repositorio
!git clone https://github.com/ChronosXV/Ingenieria_Software.git

Cloning into 'Ingenieria_Software'...
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 12 (delta 1), reused 9 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (12/12), 9.59 MiB | 13.33 MiB/s, done.
Resolving deltas: 100% (1/1), done.


In [6]:
# Me posiciono sobre el repositorio para comenzar a usar comandos git
current_dir = os.getcwd()
repo_path = 'Ingenieria_Software'

# Verificar si ya estamos en el directorio
if os.path.basename(current_dir) == repo_path:
    print(f"Ya estamos en el directorio: {current_dir}")
elif os.path.exists(repo_path) and os.path.isdir(repo_path):
    os.chdir(repo_path)
    print(f"Cambiado al directorio: {os.getcwd()}")
else:
    print(f"El directorio {repo_path} no existe o no se puede acceder.")

Cambiado al directorio: /content/Ingenieria_Software


In [9]:
# Nos posicionamos en nuestra branch
!git checkout "{branch_name_ing}"
# Validamos que estemos en nuestra branch
!git branch

Already on 'Ing_Software'
Your branch is up to date with 'origin/Ing_Software'.
* Ing_Software
  main


In [10]:
# Creamos el df con los datos de nuestro dataSet para poder analizar los datos
df_delitos = pd.read_excel('/content/Ingenieria_Software/Proyecto/DataSet_Original/delitos_2025.xlsx')


In [11]:
# Descripcion de la estructura del dataset:
print("\nEstructura del Dataset")
print(f"El dataset tiene {df_delitos.shape[0]} filas y {df_delitos.shape[1]} columnas.")


Estructura del Dataset
El dataset tiene 133203 filas y 15 columnas.


In [12]:
# Mostramos las Primeras 5 filas de nuestro DataFrame
display(df_delitos.head())

,id-mapa,anio,mes,dia,fecha,franja,tipo,subtipo,uso_arma,uso_moto,barrio,comuna,latitud,longitud,cantidad
0,1591961,2025,FEBRERO,DOMINGO,2025-02-02,14.0,Homicidios,Homicidios dolosos,NO,NO,CABALLITO,6.0,-3461534.0,-58431076.0,1
1,1591962,2025,ENERO,DOMINGO,2025-01-12,1.0,Homicidios,Homicidios dolosos,NO,NO,RETIRO,1.0,-34583965.0,-58379769.0,1
2,1591963,2025,DICIEMBRE,MIERCOLES,2025-12-10,4.0,Homicidios,Homicidios dolosos,NO,NO,BARRACAS,4.0,-34650457.0,-58404647.0,1
3,1591964,2025,ENERO,LUNES,2025-01-20,0.0,Homicidios,Homicidios dolosos,NO,NO,VILLA LUGANO,8.0,-34682057.0,-58488847.0,1
4,1591965,2025,DICIEMBRE,VIERNES,2025-12-12,23.0,Homicidios,Homicidios dolosos,NO,NO,MONSERRAT,1.0,-34608938.0,-58387545.0,1


In [13]:
# Verificamos el tipo de dato de nuestras variables
df_delitos.dtypes

,0
id-mapa,int64
anio,int64
mes,object
dia,object
fecha,datetime64[ns]
franja,float64
tipo,object
subtipo,object
uso_arma,object
uso_moto,object


Clasificación de Variables y Tipos de Datos.

Tras la carga inicial del dataset, realizamos una categorización técnica de las variables para definir la estrategia de procesamiento y análisis posterior:

1. Variables Cuantitativas (Numéricas)
Estas serán la base para cálculos estadísticos y georreferenciación.

        * Identificadores: id-mapa
        * Geolocalización: latitud, longitud
        * Métricas: cantidad
        * Atributos de Orden: anio, franja, comuna

2. Variables Cualitativas (Categóricas)
Con estos datos podremos encontrar patrones de comportamiento.

        * Descriptivas: tipo, subtipo, barrio
        * Operativas: uso_arma, uso_moto
        * Temporales: mes, dia

3. Variables de Objeto (Temporal)

        * fecha: Se identifica como la variable raíz para la cual podremos realizar análisis de series temporales, permitiendo una granularidad diaria en la evolución delictiva.

In [15]:
# Cantidad de datos nulos por columna
df_delitos.isnull().sum()

,0
id-mapa,0
anio,0
mes,0
dia,0
fecha,0
franja,155
tipo,0
subtipo,0
uso_arma,0
uso_moto,0


El conjunto de datos presenta valores nulos significativos en varias columnas. Se observa que 'franja' tiene 155 valores faltantes, mientras que 'barrio' y 'comuna' carecen de 2785 valores cada una. Además, 'latitud' y 'longitud' tienen 265 entradas nulas

In [16]:
# Realizar un análisis preliminar de calidad: porcentaje de datos faltantes por variable, detección de posibles inconsistencias.

nulos_totales = df_delitos.isnull().sum()
porcentaje_nulos = (nulos_totales / len(df_delitos)) * 100

# Armamos un cuadro resumen
calidad_df = pd.DataFrame({
    'Columna': nulos_totales.index,
    'Cantidad de Nulos': nulos_totales.values,
    'Porcentaje (%)': porcentaje_nulos.values
}).sort_values(by='Porcentaje (%)', ascending=False)

# Mostramos el resumen
print("Análisis de datos faltantes por variable:")
display(calidad_df)

Análisis de datos faltantes por variable:


,Columna,Cantidad de Nulos,Porcentaje (%)
10,barrio,2785,2.090794
11,comuna,2785,2.090794
12,latitud,265,0.198944
13,longitud,265,0.198944
5,franja,155,0.116364
1,anio,0,0.000000
0,id-mapa,0,0.000000
6,tipo,0,0.000000
4,fecha,0,0.000000
3,dia,0,0.000000


Podemos observar porcentajes bajos de nulos. Con mayores valores encontramos a las variables 'barrio' y 'comuna' con un porcentaje de (2.09%), seguidas por 'latitud' y 'longitud' (0.19%), y por ultimo 'franja' (0.11%).

In [17]:
# Verificamos si tenemos datos duplicados
print(f"Cantidad de filas duplicadas: {df_delitos.duplicated().sum()}")

Cantidad de filas duplicadas: 0


No se observan datos duplicados en el dataSet

In [20]:
# Verificacion de variable Subtipo
print("Cantidad de delitos por tipo y subtipo:")
display(df_delitos.groupby(['tipo','subtipo']).size().sort_values(ascending=False))

Cantidad de delitos por tipo y subtipo:


,,0
tipo,subtipo,
Robo,Robo total,49545
Hurto,Hurto total,45102
Vialidad,Lesiones por siniestros viales,11669
Lesiones,Lesiones Dolosas,11544
Amenazas,Amenazas,10111
Hurto,Hurto automotor,4539
Robo,Robo automotor,524
Vialidad,Muertes por siniestros viales,91
Homicidios,Homicidios dolosos,78


##Fase #3: Preparación de los Datos / Data Preparation
La fase de preparación de datos abarca todas las actividades para construir el conjunto dedatos final (datos que se introducirán en las herramientas de modelado) a partir de los datos brutos iniciales. Es probable que las tareas de preparación de datos se realicen varias veces y no en ningún orden prescrito. Las tareas incluyen la selección de tablas, registros y atributos, así como la transformación y limpieza de datos para herramientas de modelado.
* Enumerá y describí las variables relevantes que son de tu interés para el estudio.Por ejemplo, edad: es la edad en años cumplidos al 31 de diciembre de 2024, es una variable continua discreta.
* Armá una tabla que resuma el análisis estadístico descriptivo y los gráficos que realizarás para entender las características de los datos.

In [ ]:
#Acá debés cargar las librerías con las que trabajarás en esta fase



In [ ]:
#Acá debés poder realizar un EDA de los datos




##Fase #4: Modelado / Modeling
En esta fase, se seleccionan y aplican diversas técnicas de modelado y se calibran sus parámetros a valores óptimos. Normalmente, hay varias técnicas para el mismo tipo de problema de minería de datos. Algunas técnicas tienen requisitos específicos en cuanto a la forma de los datos. Por lo tanto, a menudo es necesario volver a la fase de preparación dedatos.
* Seleccioná una variable objetivo que te permita cumplir el objetivo del estudio.
* Investigá un algoritmo de Ciencia de Datos que pueda ayudarte a predecir el comportamiento de esa variable. Explicá brevemente por qué te sería útil.
* Implementa el algoritmo.

In [ ]:
#Acá debés cargar las librerías con las que trabajarás en esta fase



In [ ]:
#Acá debés poder implementar un modelo y entrenarlo




##Fase #5: Evaluación / Evaluation
En esta etapa del proyecto, ha construido un modelo (o modelos) que parece tener alta calidad desde una perspectiva de análisis de datos. Antes de proceder a la implementación final del modelo, es importante evaluar más a fondo el modelo y revisar los pasos realizados para construir el modelo y asegurarse de que logra correctamente los objetivos comerciales. Un objetivo clave es determinar si hay algún tema importante del negocio que no se hayaconsiderado lo suficiente. Al final de esta fase, se debe tomar una decisión sobre el uso de los resultados de la minería de datos.
* Investigá sobre las métricas que se utilizan para medir el modelo escogido.
* Evalúa dichas métricas en el modelo implementado e intenta explicarlas.

In [ ]:
#Acá debés cargar las librerías con las que trabajarás en esta fase



In [ ]:
#Acá debés poder validar el modelo, calcular las métricas
#y construir la matriz de confusión




##Fase #6: Despliegue (Deployment)
Por lo general, la creación del modelo no es el final del proyecto. Incluso si el propósito delmodelo es aumentar el conocimiento de los datos, el conocimiento adquirido deberá organizarse y presentarse de manera que el cliente pueda utilizarlo. A menudo implica laaplicación de modelos "en vivo" dentro de los procesos de toma de decisiones de una organización, por ejemplo, en la personalización en tiempo real de páginas web o la puntuación repetida de bases de datos de marketing. Sin embargo, en función de los requisitos, la fase de implementación puede ser tan sencilla como generar un informe o tan compleja como implementar un proceso de minería de datos repetible en toda la empresa. En muchos casos, es el cliente, no el analista de datos, quien lleva a cabo los pasos de implementación. Sin embargo, incluso si el analista no llevará a cabo el esfuerzo de implementación, es importante que el cliente comprenda de antemano qué acciones deben llevarse a cabo para hacer uso de los modelos creados.
* Definí un plan para mantener la solución actualizada y funcional. Proporcioná actualizaciones de seguridad y soluciona problemas según sea necesario.
* Colocá acá las vistas del software implementado y el link desde dónde se puede acceder para interactuar con el software (puede estar hardcodeado con las soluciones y gráficos que coloquen acá, en esta notebook).

###Aclaraciones:
* Buscá la síntesis. No nos interesan las largas explicaciones, sino la capacidad de sintetizar el problema.
* No te extiendas más de lo necesario. Utiliza tablas de resumen, herramientas gráficas. Sé creativo. Este no es un trabajo de conocimiento, es un trabajo de aplicación y comunicación de la estrategia.
* No hay respuestas incorrectas, solo problemas de negocio poco explorados 🙂